### DISCUSSION TITLE
Physical/Virtual memory on x86 kernels OHCI

### DISCUSSION THREAD

**Discussion by Ryzee119** [hathach/tinyusb#1490] [General]
I have been porting TinyUSB to an x86 win32-ish kernel. This hardware uses OHCI usb hardware.
It is working quite well, however the main headache was translating from virtual to physical memory and back. These higher level kernels cannot directly access physical memory, but the hardware must DMA from the physical address.

I have just added a couple small functions that convert a virtual memory address to a physical on and placed them in the relevant locations in https://github.com/hathach/tinyusb/blob/master/src/portable/ohci/ohci.c (TD/ED linking, hcca, list heads etc that HW accesses directly.).

I added two functions like this (maybe we could use weak functions that just return the original address on embedded hardware?):

`return void *phys_addr(void *virtual_address)` and
`return void *virt_addr(void *physical_address)`
It is currently used like this:
`OHCI_REG->hcca            = (uint32_t) phys_addr(&ohci_data.hcca);`

I can raise a PR to upsteam these changes but just wanting some ideas or suggestions if theres a better way before doing so.

**Comment by hathach** [comment]
yeah, please make an PR if possible, I have written this OHCI driver a long time ago for NXP LPC MCU and didn't take care of the virtual/physical and/or mem caching. Just make it as abstract as you could, I will pull out my old nxp board to test with and help to make the changes if needed.

**Reply by Ryzee119** [reply]
The relevant commit is https://github.com/Ryzee119/tinyusb/commit/2004e343ba317faa38fcc5012fae6be39e8d7c77
Can confirm it is working on my end which requires this differentiation.
